File: src/main/java/com/warrantybee/api/helpers/HashHelper.java
package com.warrantybee.api.helpers;

import de.mkammerer.argon2.Argon2;
import de.mkammerer.argon2.Argon2Factory;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;

/**
 * Helper class providing secure hashing and verification methods
 * using the Argon2id algorithm for password or sensitive data protection.
 */
public class HashHelper {

    /** Argon2 instance configured with default parameters for hashing and verification. */
    private static final Argon2 argon2 = Argon2Factory.create(
            Argon2Factory.Argon2Types.ARGON2id, 32, 64
    );

    /** Number of Argon2 iterations to perform during hashing. */
    private static final int _ITERATIONS = 3;

    /** Memory cost (in kilobytes) used by the Argon2 algorithm. */
    private static final int _MEMORY = 65536;

    /** Degree of parallelism (number of threads) used for hashing. */
    private static final int _PARALLELISM = 1;

    /**
     * Generates a secure Argon2 hash for the specified plain text input.
     *
     * @param text the plain text to hash.
     * @return the generated Argon2 hash as a string.
     * @throws IllegalArgumentException if the input text is {@code null}.
     */
    public static String getHash(String text) {
        if (text == null) {
            throw new IllegalArgumentException("Password cannot be null");
        }

        String argonHash = argon2.hash(_ITERATIONS, _MEMORY, _PARALLELISM, text.getBytes(StandardCharsets.UTF_8));
        return _encodeBase64(argonHash);
    }

    /**
     * Validates a plain text input against a previously generated Argon2 hash.
     * The comparison is performed in a time-constant manner to prevent timing attacks.
     *
     * @param text the plain text to verify.
     * @param storedHash the Argon2 hash to compare against.
     * @return {@code true} if the text matches the stored hash, {@code false} otherwise.
     */
    public static boolean verify(String text, String storedHash) {
        if (text == null || storedHash == null) {
            return false;
        }

        String decodedHash = _decodeBase64(storedHash);
        return argon2.verify(decodedHash, text.getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Generates a random token using SHA256.
     *
     * @return A random token.
     */
    public static String generateToken() {
        try {
            SecureRandom secureRandom = new SecureRandom();
            byte[] randomBytes = new byte[32];
            secureRandom.nextBytes(randomBytes);

            MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
            byte[] hashedBytes = messageDigest.digest(randomBytes);

            return _bytesToHex(hashedBytes);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 algorithm not found", e);
        }
    }

    /**
     * Converts a byte array into its hexadecimal string representation.
     * Each byte is represented by two hexadecimal characters.
     *
     * @param bytes the byte array to convert
     * @return a hexadecimal string representation of the given bytes
     */
    private static String _bytesToHex(byte[] bytes) {
        StringBuilder hexString = new StringBuilder(2 * bytes.length);
        for (byte b : bytes) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }
        return hexString.toString();
    }

    /**
     * Encodes the given string into a Base64-encoded string.
     *
     * @param value the plain text to encode
     * @return the Base64-encoded representation of the input
     */
    private static String _encodeBase64(String value) {
        return Base64.getEncoder().encodeToString(value.getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Decodes the given Base64-encoded string back to its original text.
     *
     * @param base64 the Base64 string to decode
     * @return the decoded plain text
     */
    private static String _decodeBase64(String base64) {
        byte[] decoded = Base64.getDecoder().decode(base64);
        return new String(decoded, StandardCharsets.UTF_8);
    }
}

File: src/WarrantyBee.Application/Common/HashHelper.cs
using System.Security.Cryptography;
using System.Text;
using Konscious.Security.Cryptography;

namespace WarrantyBee.Application.Common;

/// <summary>
/// Provides utility methods for hashing and verifying passwords and tokens using Argon2id.
/// </summary>
public static class HashHelper
{
    private const int DefaultIterations = 3;
    private const int DefaultMemory = 65536;
    private const int DefaultParallelism = 1;

    /// <summary>
    /// Generates a secure Argon2id hash for the specified text.
    /// </summary>
    /// <param name="text">The text to hash.</param>
    /// <returns>A standard Argon2 string: $argon2id$v=19$m=65536,t=3,p=1$salt$hash</returns>
    /// <exception cref="ArgumentNullException">Thrown when text is null.</exception>
    public static string GetHash(string text)
    {
        if (text == null) throw new ArgumentNullException(nameof(text));

        var salt = new byte[16];
        RandomNumberGenerator.Fill(salt);

        using var argon2 = new Argon2id(Encoding.UTF8.GetBytes(text));
        argon2.Salt = salt;
        argon2.DegreeOfParallelism = DefaultParallelism;
        argon2.MemorySize = DefaultMemory;
        argon2.Iterations = DefaultIterations;

        var hash = argon2.GetBytes(32);

        // Return a standard Argon2 string format
        return $"$argon2id$v=19$m={DefaultMemory},t={DefaultIterations},p={DefaultParallelism}${Convert.ToBase64String(salt)}${Convert.ToBase64String(hash)}";
    }

    /// <summary>
    /// Verifies a text against a stored Argon2id hash. Supports standard Argon2 strings and legacy wrapped formats.
    /// </summary>
    /// <param name="text">The text to verify.</param>
    /// <param name="storedHash">The stored hash string.</param>
    /// <returns>True if the text matches the hash; otherwise, false.</returns>
    public static bool Verify(string text, string storedHash)
    {
        if (string.IsNullOrWhiteSpace(text) || string.IsNullOrWhiteSpace(storedHash)) return false;

        try
        {
            string rawHash = storedHash;

            // Handle legacy double-Base64 wrapping if detected
            if (!storedHash.StartsWith("$") && IsBase64String(storedHash))
            {
                try
                {
                    rawHash = Encoding.UTF8.GetString(Convert.FromBase64String(storedHash));
                }
                catch { /* Ignore and try processing as-is */ }
            }

            var parts = rawHash.Split('$');
            if (parts.Length < 6) return false;

            // Extract parameters: m=65536,t=3,p=1
            var paramsPart = parts[3];
            int memory = DefaultMemory;
            int iterations = DefaultIterations;
            int parallelism = DefaultParallelism;

            foreach (var p in paramsPart.Split(','))
            {
                var kv = p.Split('=');
                if (kv.Length != 2) continue;
                switch (kv[0])
                {
                    case "m": memory = int.Parse(kv[1]); break;
                    case "t": iterations = int.Parse(kv[1]); break;
                    case "p": parallelism = int.Parse(kv[1]); break;
                }
            }

            var salt = Convert.FromBase64String(parts[4]);
            var hash = Convert.FromBase64String(parts[5]);

            using var argon2 = new Argon2id(Encoding.UTF8.GetBytes(text));
            argon2.Salt = salt;
            argon2.DegreeOfParallelism = parallelism;
            argon2.MemorySize = memory;
            argon2.Iterations = iterations;

            var newHash = argon2.GetBytes(32);
            return CryptographicOperations.FixedTimeEquals(hash, newHash);
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Generates a secure random token.
    /// </summary>
    /// <returns>A hexadecimal string representation of the generated token.</returns>
    public static string GenerateToken()
    {
        var randomBytes = new byte[32];
        RandomNumberGenerator.Fill(randomBytes);
        using var sha256 = SHA256.Create();
        var hashedBytes = sha256.ComputeHash(randomBytes);
        return Convert.ToHexString(hashedBytes).ToLower();
    }

    private static bool IsBase64String(string s)
    {
        if (string.IsNullOrWhiteSpace(s)) return false;
        if (s.Length % 4 != 0) return false;
        try
        {
            Convert.FromBase64String(s);
            return true;
        }
        catch
        {
            return false;
        }
    }
}
